### PR TITLE
[EXP-33-234] refactor: curve oracle and curve v5 support

### LINKED ISSUE #213: [EXP-11-213] Support curve tokens v5.0.0
token [https://etherscan.io/address/0x29059568bb40344487d62f7450e78b8e6c74e0e5](https://etherscan.io/address/0x29059568bb40344487d62f7450e78b8e6c74e0e5)\
pool [https://etherscan.io/address/0xc26b89a667578ec7b3f11b2f98d6fd15c07c54ba](https://etherscan.io/address/0xc26b89a667578ec7b3f11b2f98d6fd15c07c54ba)

creation tx https://ethtx.info/mainnet/0x9718eb7885e31ac87e72ecd57cb826438574295f25c19db17988687fab1445fd/
factory https://etherscan.io/address/0xf18056bbd320e96a48e3fbf8bc061322531aac99#code

### PR DESCRIPTION
**Pull request by banteg**
* add support for curve v5 factory (fixes #213)
* build lp token to pool mapping on load (`get_lp_token`, `get_token`)
* add cached mappings from registries/factories to pools
* add address provider ids enum for readability
* watch for curve registry updates
* watch for new registries deployed
* remove and generalize cryptoswap registry logic
* deprecate `registry` in favor of `get_registry(pool)`
* change `metapools_by_factory` to cached `factories`
* convert token and pool args to checksum address
* add `read_pool` utility to fetch pool_list/pool_count type registries commonly found in curve
* vastly simplify logic around factory and registry api

### CODE REVIEW

**Comment by BobTheBuidler** [comment]
LGTM, though there's a lot going on here so I'm going to wait for another pair of eyes before merging.

**Review by mariuspod — APPROVED** [review]
Added some changes and tested on staging.
@banteg in curve#get_underlying_coins I've added another ugly `hasattr` check, not sure if there's a better way.

`curve.get_underlying_coins('0xC26b89A667578ec7b3f11b2F98d6Fd15C07C54ba')` fails without this

**Pull request merged**
